src/lib.rs

```diff
@@ -57,66 +63,129 @@ pub fn dummy_endpoint() -> String {
+///     endpoint.incoming(core.handle())
+///         .expect("failed to open a pipe")
+///         .for_each(|(stream, _)| {
+///             println!("Connection received");
+///             future::ok(())
 ///     });
 /// }
 /// ```
 pub struct Endpoint {
-    _path: String,
-    _handle: Handle,
-    #[cfg(not(windows))]
-    inner: tokio_uds::UnixListener,
+    path: String,
     #[cfg(windows)]
-    inner: NamedPipe,
+    security_attributes: *mut SECURITY_ATTRIBUTES,
 }
 
 impl Endpoint {
     /// Stream of incoming connections
     #[cfg(not(windows))]
-    pub fn incoming(self) -> Incoming {
-        Incoming { inner: self.inner.incoming() }
+    pub fn incoming(self, handle: Handle) -> io::Result<Incoming> {
+        Ok(
+            Incoming { inner: self.inner(&handle)?.incoming() }
+          )
     }
 
     /// Stream of incoming connections    
     #[cfg(windows)]
-    pub fn incoming(self) -> Incoming {
-        Incoming { inner: NamedPipeSupport { path: self._path, handle: self._handle.remote().clone(), pipe: self.inner } }
+    pub fn incoming(self, handle: Handle) -> io::Result<Incoming> {
+        let pipe = self.inner(&handle)?;
+        Ok(
+            Incoming { inner: NamedPipeSupport { path: self.path, handle: handle.remote().clone(), pipe: pipe, security_attributes: self.security_attributes} }
+          )
     }
 
     /// Inner platform-dependant state of the endpoint
     #[cfg(windows)]
-    fn inner(p: &str, handle: &Handle) -> io::Result<NamedPipe> {
-        NamedPipe::new(p, handle)
+    fn inner(&self, handle: &Handle) -> io::Result<NamedPipe> {
+        extern crate mio_named_pipes;
+        use std::os::windows::io::*;
+        use miow::pipe::NamedPipeBuilder;
+
+        let raw_handle = unsafe { NamedPipeBuilder::new(&self.path)
+            .first(true)
+            .inbound(true)
+            .outbound(true)
+            .out_buffer_size(65536)
+            .in_buffer_size(65536)
+            .with_security_attributes(self.security_attributes)?
+            .into_raw_handle()};
+
+        let mio_pipe = unsafe { mio_named_pipes::NamedPipe::from_raw_handle(raw_handle) };
+        NamedPipe::from_pipe(mio_pipe, handle)
     }
 
     /// Inner platform-dependant state of the endpoint
     #[cfg(not(windows))]
-    fn inner(p: &str, handle: &Handle) -> io::Result<tokio_uds::UnixListener> {
-        tokio_uds::UnixListener::bind(p, handle)
+    fn inner(&self, handle: &Handle) -> io::Result<tokio_uds::UnixListener> {
+        tokio_uds::UnixListener::bind(&self.path, handle)
     }
 
+    /// Sets the security attributes of the underlying named pipes. Note that if the pointer to the
+    /// security attributes is not null, it should be valid for the lifetime of this struct and the
+    /// lifetime of the returned `Incoming` struct.
+    #[cfg(windows)]
+    pub unsafe fn set_security_attributes(&mut self, security_attributes: *mut SECURITY_ATTRIBUTES) {
```

Comment:
does it need to be public?

Reply:
There has to be _some_ way to set the pointer to the security attributes. It could be set in the `new()` function as well, but I that would make it unsafe as well. I'm open to suggestions about better ways to pass the pointer around.

Reply:
I think it is related to the suggestion below

There should be some kind of safe wrapper, and public api should expose only operations using this safe wrapper

Reply:
Oh, you mean a safe wrapper for `SECURITY_ATTRIBUTES`? Currently, there are no safe libraries that wrap the Windows NT ACL API and implementing a useful one is not trivial, although we have considered doing so. For now, we're trying to get away with exposing this as an unsafe function that takes a pointer.